microsoft/PowerToys
# [PT Run] Improvements on EnvironmentHelper and deletion of old env vars

## Summary of the Pull Request

**What is this about:**
While adding exception handling to `EnvironmentHelper` (other PR) I detected that removed machine and user environment variables are not deleted when updating the process environment.

**What is include in the PR:** 
- Improving the code.
- Fix: Path variable was merged incorrectly.
- Added: Feature to delete variables that doesn't exist anymore.
- Handle variable name casing correctly.
![image](https://user-images.githubusercontent.com/61519853/138588863-15a09fe8-4286-41b5-a5e5-df6504578e12.png)


**ToDo-List**
- [x] Code optimization: var names like `uVar` and `mVar`
- [x] Case handling for variable names
   - [x] Test what happens in windows/cmd if `Path` variable exists with different var name casing (Path != PATH) for user and machine.
   - [x] Analyze windows' behavior if we have a general variable that has different casing (name vs Name vs NAME) on user and machine level
   - [x] Making dict and hashSet case insensitive that machine key casing is used .
   - [x] add code to correct casing on case change.
- [x] Error handling where we get or set env vars and duration cost logging
- [x] Add code to remove unchanged vars from dictionary
- [x] Add else statement to log skipped vars if listed as protected
- [x] fix merge conflicts

**How does someone test / validate:** 
I plan to test the code manually. The following test are planned:
- [x] Protected vars list
   - List gets generated: **success**
   - List content is correct: **success**
   - Vars on the list can't be changed on update: **success**
- [x] Machine vars
   - add new variable: **success**
   - change value: **success**
   - change name: **success**
   - delete: **success**
- [x] User vars
   - add new variable: **success**
   - change value: **success**
   - change name: **success**
   - delete: **success**
- [x] Merge user and machine vars
   - Overwrite machine with user var: **success**
   - Delete user, keep machine: **success**
   - delete machine, keep user: **success**
   - delete both: **success**
- [x] Path variable
   - User (add, change, remove): **success**
   - Machine (add, change, remove): **success**
   - Merge User and machine with same path snipped: **success**
- [x] Special tests
   - Change casing of name (testVariable -> Testvariable): **success**
   - Check if delete code (line 98) is case sensitive: **success**

## Quality Checklist

- [x] **Linked issue:** #13206

## Contributor License Agreement (CLA)
A CLA must be signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/PowerToys) and sign the CLA.
